pytential/linalg/cluster.py

```diff
@@ -0,0 +1,397 @@
+__copyright__ = "Copyright (C) 2022 Alexandru Fikl"
+
+__license__ = """
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in
+all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+"""
+
+from dataclasses import dataclass, replace
+from functools import singledispatch
+from typing import Any, Iterator, Optional, Tuple
+
+import numpy as np
+
+from pytools import memoize_method
+from pytools.obj_array import make_obj_array
+
+from arraycontext import PyOpenCLArrayContext
+from boxtree.tree import Tree
+from meshmode.discretization import Discretization
+from pytential import sym, GeometryCollection
+from pytential.linalg.utils import IndexList, TargetAndSourceClusterList
+from pytential.qbx import QBXLayerPotentialSource
+
+__doc__ = """
+Clustering
+~~~~~~~~~~
+
+.. autoclass:: ClusterLevel
+.. autoclass:: ClusterTree
+
+.. autofunction:: split_array
+.. autofunction:: cluster
+.. autofunction:: uncluster
+
+.. autofunction:: partition_by_nodes
+"""
+
+# FIXME: this is just an arbitrary value
+_DEFAULT_MAX_PARTICLES_IN_BOX = 32
+
+
+# {{{ cluster tree
+
+
+def make_cluster_parent_map(parent_ids: np.ndarray) -> np.ndarray:
+    """Construct a parent map for :attr:`ClusterLevel.parent_map`."""
```

Comment:
Rephrase/rework? "All children in a bucket"?

Reply:
Rephrased and added some more context in https://github.com/inducer/pytential/compare/e9643cc480e53575274e4dc8e14d81d9b060f6b6..20fbc31a57c1a9a642abfa0a9b7a83ad81f1a1d5. Thoughts?